### ISSUE TITLE
AssertionError indicating that 'StartUtteranceBotAction' events must include a 'script' of type 'str'. with the version 0.8.0

### ISSUE THREAD

**Issue by praveen062** [NVIDIA/NeMo-Guardrails#385]
Upon posing several questions, an error arises: AssertionError indicating that 'StartUtteranceBotAction' events must include a 'script' of type 'str'. While the function 'rails.generate_async(messages=history)' executes successfully for initial queries, it fails after a certain number of queries with the aforementioned AssertionError. This issue is observed with version tag 0.8.0.

**Comment by drazvan** [comment]
Can you provide a minimal example of how to reproduce? Can you reproduce it on the `/examples/bots/abc` configuration?

**Comment by praveen062** [comment]
https://github.com/NVIDIA/NeMo-Guardrails/assets/5464774/b9bdfc02-3e45-4f7a-91a2-021a48357964

https://github.com/praveen062/NeMo-Guardrails/blob/ExampleForChatHistoryIssue/examples/bots/abc/example.ipynb

**Comment by drazvan** [comment]
Thanks! Based on this I think there's something happening when the conversation history gets caped. We need to investigate this.

**Comment by Wildshire** [comment]
Not sure if I can help debugging this but, I did some tests between `0.7.1` and `0.8.0` and these are my findings.

**Note**
Let's put `history1` as

```
[
  {
    "role": "user",
    "content": "question"
  }
]
```

and `history2` as:

```
[
  {
    "role": "user",
    "content": "question"
  },
  {
    "role": "assistant",
    "content": "reponse"
  },
  {
    "role": "user",
    "content": "question"
  }
]
```

On `0.7.1`
- After startup (a.k.a -> first generation), you can pass `history1` to `rails.generate_async` without any issues 
- After startup (a.k.a -> first generation), you can pass `history2` to `rails.generate_async` without any issues
- After multple generations (a.k.a -> you have called `rails.generate_async` multiple times), you can pass `history1` to `rails.generate_async` without any issues 
- After multple generations (a.k.a -> you have called `rails.generate_async` multiple times), you can pass `history2` to `rails.generate_async` without any issues 

On `0.8.0`
- After startup (a.k.a -> first generation), you can pass `history1` to `rails.generate_async` without any issues
- After startup (a.k.a -> first generation), you CANNOT pass `history2` to `rails.generate_async` without any issues.  -> `AssertionError indicating that 'StartUtteranceBotAction' events must include a 'script' of type 'str'`
- After FIRST generation (a.k.a -> you used `history1` after startup), you CAN pass `history2` to `rails.generate_async` WITHOUT any issues. This means that if you generate again, `history` should have 5 messages and so on.

For me this kinda means that perhaps on `0.8.0` you have to be consistent with the calls and the internal history that NeMo handles, but I prefer having total freedom like on `0.7.1` because then you can generate any situation presented to the system.

If more context is needed or this is not related, let me know.

Best

**Comment by drazvan** [comment]
#392 is now merged and should fix the issue. Indeed, when the full history (e.g., `history_2`) was used directly, and no previous intermediate steps were found, there was an error when constructing the equivalent stream of events. Thanks @saradiazdelser!

**Comment by Wildshire** [comment]
Hello

I am experiencing this error again in version 0.8.x. Is is just me or can someone confirm this?

Best

**Comment by drazvan** [comment]
Unfortunately, the fix got undone in a merge. It's now fixed again in #454. We'll publish a patch release (0.8.3) this week which will fix this, again.
